File: ui/client/src/api/schedules.ts
import axios from '@/http/axiosAgentConfig';

// ────────────────────────────────────────────────────────────────────────────────
// Types
// ────────────────────────────────────────────────────────────────────────────────

export interface ScheduleDefinitionInput {
  interval?: string;
  cron_expression?: string;
  overlap_policy?: string;
  timezone?: string;
  start_at?: string;
  end_at?: string | null;
  remaining_actions?: number | null;
}

export interface CreateScheduleInput {
  blueprintId: string;
  inputs: { user_prompt: string } & Record<string, unknown>;
  source?: string;
  schedule: ScheduleDefinitionInput;
}

export interface UpdateScheduleInput {
  scheduleId: string;
  inputs?: { user_prompt?: string } & Record<string, unknown>;
  schedule?: ScheduleDefinitionInput;
}

export interface RunStatusEntry {
  session_id: string;
  status: string;
  started_at?: string;
}

export interface RunStats {
  total_runs: number;
  last_run_at?: string | null;
  recent_statuses: RunStatusEntry[];
}

export interface WorkflowScheduleResponse {
  id: string;
  blueprint_id: string;
  blueprint_name?: string;
  identity: { type: string; id: string; display_name: string };
  inputs: { user_prompt: string } & Record<string, unknown>;
  source: string;
  schedule: ScheduleDefinitionInput;
  schedule_status: string;
  engine_handle?: string;
  completed_at?: string;
  run_stats: RunStats;
}

export interface ScheduleRunResponse {
  session_id: string;
  status: string;
  started_at: string;
  metadata: Record<string, unknown>;
}

function withTeamScope<T extends Record<string, unknown>>(
  payload: T,
  teamId?: string,
): T & { teamId?: string } {
  if (teamId) return { ...payload, teamId };
  return payload;
}

// ────────────────────────────────────────────────────────────────────────────────
// CRUD Operations
// ────────────────────────────────────────────────────────────────────────────────

export async function createSchedule(
  input: CreateScheduleInput,
  teamId?: string,
): Promise<WorkflowScheduleResponse> {
  const { data } = await axios.post(
    '/schedules/schedule.create',
    withTeamScope({ ...input }, teamId),
  );
  return data;
}

export async function updateSchedule(
  input: UpdateScheduleInput,
  teamId?: string,
): Promise<WorkflowScheduleResponse> {
  const { data } = await axios.post(
    '/schedules/schedule.update',
    withTeamScope({ ...input }, teamId),
  );
  return data;
}

export async function listSchedules(
  teamId?: string,
  blueprintId?: string,
): Promise<WorkflowScheduleResponse[]> {
  const params: Record<string, string> = {};
  if (teamId) params.teamId = teamId;
  if (blueprintId) params.blueprintId = blueprintId;
  const { data } = await axios.get('/schedules/schedule.list', { params });
  return data;
}

export async function getSchedule(
  scheduleId: string,
  teamId?: string,
): Promise<WorkflowScheduleResponse> {
  const params: Record<string, string> = { scheduleId };
  if (teamId) params.teamId = teamId;
  const { data } = await axios.get('/schedules/schedule.get', { params });
  return data;
}

// ────────────────────────────────────────────────────────────────────────────────
// Schedule Lifecycle
// ────────────────────────────────────────────────────────────────────────────────

export async function pauseSchedule(
  scheduleId: string,
  teamId?: string,
): Promise<WorkflowScheduleResponse> {
  const { data } = await axios.post(
    '/schedules/schedule.pause',
    withTeamScope({ scheduleId }, teamId),
  );
  return data;
}

export async function resumeSchedule(
  scheduleId: string,
  teamId?: string,
): Promise<WorkflowScheduleResponse> {
  const { data } = await axios.post(
    '/schedules/schedule.resume',
    withTeamScope({ scheduleId }, teamId),
  );
  return data;
}

export async function triggerSchedule(
  scheduleId: string,
  teamId?: string,
): Promise<WorkflowScheduleResponse> {
  const { data } = await axios.post(
    '/schedules/schedule.trigger',
    withTeamScope({ scheduleId }, teamId),
  );
  return data;
}

export async function deleteSchedule(
  scheduleId: string,
  teamId?: string,
): Promise<{ deleted: boolean }> {
  const { data } = await axios.delete('/schedules/schedule.delete', {
    data: withTeamScope({ scheduleId }, teamId),
  });
  return data;
}

// ────────────────────────────────────────────────────────────────────────────────
// Schedule Runs (session history for a workflow schedule)
// ────────────────────────────────────────────────────────────────────────────────

export async function getScheduleRuns(
  scheduleId: string,
  teamId?: string,
  limit?: number,
): Promise<ScheduleRunResponse[]> {
  const params: Record<string, string | number> = { scheduleId };
  if (teamId) params.teamId = teamId;
  if (limit) params.limit = limit;
  const { data } = await axios.get('/schedules/schedule.runs', { params });
  return data;
}

File: multi-agent/adapters/inbound/flask/endpoints/schedules.py
"""
Flask endpoints for workflow schedules.

Provides CRUD for WorkflowSchedule entities and schedule lifecycle
operations (pause, resume, trigger, delete). All endpoints are identity-scoped.
"""
import logging

from flask import Blueprint, jsonify, current_app, g
from global_utils.helpers.apiargs import from_body, from_query
from webargs import fields

from mas.scheduling.service import (
    ScheduleLimitExceededError,
    ScheduleNotFoundError,
    SchedulePermissionError,
)
from mas.blueprints.exceptions import BlueprintNotFoundError
from inbound.flask.decorators import with_require_identity_authorization

logger = logging.getLogger(__name__)

_RESPONSE_EXCLUDE = {"credential_user_id"}

schedules_bp = Blueprint("schedules", __name__)


@schedules_bp.route("/schedule.create", methods=["POST"])
@with_require_identity_authorization
@from_body({
    "blueprint_id": fields.Str(data_key="blueprintId", required=True),
    "inputs": fields.Dict(data_key="inputs", required=True),
    "source": fields.Str(data_key="source", load_default="manual"),
    "schedule": fields.Dict(data_key="schedule", required=True),
})
def create_schedule(identity, blueprint_id, inputs, source, schedule):
    try:
        svc = current_app.container.schedule_service
        wf_schedule = svc.create(
            identity=identity,
            blueprint_id=blueprint_id,
            inputs=inputs,
            source=source,
            schedule=schedule,
            credential_user_id=getattr(g, "identity_username", ""),
        )
        return jsonify(wf_schedule.model_dump(mode="json", exclude=_RESPONSE_EXCLUDE)), 201
    except BlueprintNotFoundError as e:
        return jsonify({"error": str(e), "error_type": "BLUEPRINT_NOT_FOUND"}), 404
    except SchedulePermissionError as e:
        return jsonify({"error": str(e), "error_type": "FORBIDDEN"}), 403
    except ScheduleLimitExceededError as e:
        return jsonify({"error": str(e), "error_type": "LIMIT_EXCEEDED"}), 409
    except ValueError as e:
        return jsonify({"error": str(e), "error_type": "VALIDATION_ERROR"}), 400
    except Exception:
        logger.exception("Unhandled error in create_schedule")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500


@schedules_bp.route("/schedule.update", methods=["POST"])
@with_require_identity_authorization
@from_body({
    "schedule_id": fields.Str(data_key="scheduleId", required=True),
    "inputs": fields.Dict(data_key="inputs", load_default=None),
    "schedule": fields.Dict(data_key="schedule", load_default=None),
})
def update_schedule(identity, schedule_id, inputs, schedule):
    try:
        svc = current_app.container.schedule_service
        wf_schedule = svc.update(
            schedule_id,
            identity=identity,
            inputs=inputs,
            schedule=schedule,
        )
        return jsonify(wf_schedule.model_dump(mode="json", exclude=_RESPONSE_EXCLUDE)), 200
    except ScheduleNotFoundError as e:
        return jsonify({"error": str(e), "error_type": "NOT_FOUND"}), 404
    except SchedulePermissionError as e:
        return jsonify({"error": str(e), "error_type": "FORBIDDEN"}), 403
    except ValueError as e:
        return jsonify({"error": str(e), "error_type": "VALIDATION_ERROR"}), 400
    except Exception:
        logger.exception("Unhandled error in update_schedule")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500


@schedules_bp.route("/schedule.list", methods=["GET"])
@with_require_identity_authorization
@from_query({
    "blueprint_id": fields.Str(data_key="blueprintId", load_default=None),
})
def list_schedules(identity, blueprint_id):
    try:
        svc = current_app.container.schedule_service
        result = svc.list_enriched(identity=identity, blueprint_id=blueprint_id)
        return jsonify(result), 200
    except Exception:
        logger.exception("Unhandled error in list_schedules")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500


@schedules_bp.route("/schedule.get", methods=["GET"])
@with_require_identity_authorization
@from_query({
    "schedule_id": fields.Str(data_key="scheduleId", required=True),
})
def get_schedule(identity, schedule_id):
    try:
        svc = current_app.container.schedule_service
        wf_schedule = svc.get(schedule_id, identity=identity)
        return jsonify(wf_schedule.model_dump(mode="json", exclude=_RESPONSE_EXCLUDE)), 200
    except ScheduleNotFoundError as e:
        return jsonify({"error": str(e), "error_type": "NOT_FOUND"}), 404
    except SchedulePermissionError as e:
        return jsonify({"error": str(e), "error_type": "FORBIDDEN"}), 403
    except Exception:
        logger.exception("Unhandled error in get_schedule")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500


@schedules_bp.route("/schedule.pause", methods=["POST"])
@with_require_identity_authorization
@from_body({
    "schedule_id": fields.Str(data_key="scheduleId", required=True),
})
def pause_schedule(identity, schedule_id):
    try:
        svc = current_app.container.schedule_service
        wf_schedule = svc.pause(schedule_id, identity=identity)
        return jsonify(wf_schedule.model_dump(mode="json", exclude=_RESPONSE_EXCLUDE)), 200
    except ScheduleNotFoundError as e:
        return jsonify({"error": str(e), "error_type": "NOT_FOUND"}), 404
    except SchedulePermissionError as e:
        return jsonify({"error": str(e), "error_type": "FORBIDDEN"}), 403
    except Exception:
        logger.exception("Unhandled error in pause_schedule")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500


@schedules_bp.route("/schedule.resume", methods=["POST"])
@with_require_identity_authorization
@from_body({
    "schedule_id": fields.Str(data_key="scheduleId", required=True),
})
def resume_schedule(identity, schedule_id):
    try:
        svc = current_app.container.schedule_service
        wf_schedule = svc.resume(schedule_id, identity=identity)
        return jsonify(wf_schedule.model_dump(mode="json", exclude=_RESPONSE_EXCLUDE)), 200
    except ScheduleNotFoundError as e:
        return jsonify({"error": str(e), "error_type": "NOT_FOUND"}), 404
    except SchedulePermissionError as e:
        return jsonify({"error": str(e), "error_type": "FORBIDDEN"}), 403
    except Exception:
        logger.exception("Unhandled error in resume_schedule")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500


@schedules_bp.route("/schedule.trigger", methods=["POST"])
@with_require_identity_authorization
@from_body({
    "schedule_id": fields.Str(data_key="scheduleId", required=True),
})
def trigger_schedule(identity, schedule_id):
    try:
        svc = current_app.container.schedule_service
        wf_schedule = svc.trigger(schedule_id, identity=identity)
        return jsonify(wf_schedule.model_dump(mode="json", exclude=_RESPONSE_EXCLUDE)), 200
    except ScheduleNotFoundError as e:
        return jsonify({"error": str(e), "error_type": "NOT_FOUND"}), 404
    except SchedulePermissionError as e:
        return jsonify({"error": str(e), "error_type": "FORBIDDEN"}), 403
    except ValueError as e:
        return jsonify({"error": str(e), "error_type": "VALIDATION_ERROR"}), 400
    except Exception:
        logger.exception("Unhandled error in trigger_schedule")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500


@schedules_bp.route("/schedule.delete", methods=["DELETE"])
@with_require_identity_authorization
@from_body({
    "schedule_id": fields.Str(data_key="scheduleId", required=True),
})
def delete_schedule(identity, schedule_id):
    try:
        svc = current_app.container.schedule_service
        svc.delete(schedule_id, identity=identity)
        return jsonify({"deleted": True}), 200
    except ScheduleNotFoundError as e:
        return jsonify({"error": str(e), "error_type": "NOT_FOUND"}), 404
    except SchedulePermissionError as e:
        return jsonify({"error": str(e), "error_type": "FORBIDDEN"}), 403
    except Exception:
        logger.exception("Unhandled error in delete_schedule")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500


@schedules_bp.route("/schedule.runs", methods=["GET"])
@with_require_identity_authorization
@from_query({
    "schedule_id": fields.Str(data_key="scheduleId", required=True),
    "limit": fields.Int(data_key="limit", load_default=20),
})
def get_schedule_runs(identity, schedule_id, limit):
    try:
        svc = current_app.container.schedule_service
        runs = svc.get_runs(schedule_id, identity=identity, limit=limit)
        return jsonify([r.model_dump(mode="json") for r in runs]), 200
    except ScheduleNotFoundError as e:
        return jsonify({"error": str(e), "error_type": "NOT_FOUND"}), 404
    except SchedulePermissionError as e:
        return jsonify({"error": str(e), "error_type": "FORBIDDEN"}), 403
    except Exception:
        logger.exception("Unhandled error in get_schedule_runs")
        return jsonify({"error": "Internal server error", "error_type": "INTERNAL_ERROR"}), 500
